contrib/epee/include/hex.h

```diff
@@ -0,0 +1,65 @@
+// Copyright (c) 2017, The Monero Project
+//
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without modification, are
+// permitted provided that the following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice, this list of
+//    conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list
+//    of conditions and the following disclaimer in the documentation and/or other
+//    materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be
+//    used to endorse or promote products derived from this software without specific
+//    prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#pragma once
+
+#include <array>
+#include <cstdint>
+#include <iosfwd>
+#include <string>
+
+#include "span.h"
+
+namespace epee
+{
+  struct to_hex
+  {
+    //! \return A std::string containing hex of `src`.
+    static std::string string(const span<const std::uint8_t> src);
```

Comment:
It was intentional - its going to be treated identically to separate ptr,length parameters in most (maybe even all) circumstances.

If the `span` was taken by reference, places where a temporary `span` are created require an object to be created on the stack so a memory address can be provided to the callee. Taking it by value means each field of `span` can be passed by register (depending on the ABI). The x86-64 ABI passes the internal `ptr,length` of `span` via register (I disassembled gcc output). My reading of the armeabi and x32 ABI indicates the same should happen. Older i386 ABI uses stack anyway, and no clue what Microsoft does.

So all of the `pod_to_hex` calls result in a dynamic pointer address being written to a register, and a fixed value being written to a second register. I've seen some recommendations to _always_ taking a struct whose `sizeof` exceeds `void*` by reference (this came up on the Boost ML recently). In this case it would mean one less integer "copy" per caller who already had a `span`. So the penalty is small for the currently rare case. And there's less indirection for the common case.